library/include/BitArray.h

```diff
@@ -526,27 +407,125 @@ namespace DFHack
         {
             return end();
         }
+
+        iterator erase(const_iterator pos)
+        {
+            auto root = static_cast<L *>(this);
+            CHECK_INVALID_ARGUMENT(pos.iter.root == root);
+            CHECK_NULL_POINTER(pos.iter.cur);
+
+            auto link = pos.iter.cur;
+            auto next = link->next;
+            if (link->next && link->prev)
+            {
+                link->next->prev = link->prev;
+                link->prev->next = link->next;
+            }
+            else if (link->next)
+            {
+                link->next->prev = nullptr;
+            }
+            else if (link->prev)
+            {
+                link->prev->next = nullptr;
+            }
+
+            if (root->next == link)
+            {
+                root->next = next;
+            }
+
+            delete link;
+
+            return iterator(root, next);
+        }
+        iterator insert(const_iterator pos, I * const & item)
+        {
+            auto root = static_cast<L *>(this);
+            CHECK_INVALID_ARGUMENT(pos.iter.root == root);
+
+            auto link = pos.iter.cur;
+            if (!link || !link->prev)
+            {
+                if (!link && root->next)
+                {
+                    pos--;
+                    return insert_after(pos, item);
+                }
+
+                CHECK_INVALID_ARGUMENT(root->next == link);
+                push_front(item);
+                return begin();
+            }
+
+            auto newlink = new L();
+            newlink->prev = link->prev;
+            newlink->next = link;
+            link->prev = newlink;
+            if (newlink->prev)
+            {
+                newlink->prev->next = newlink;
+            }
+            else if (link == root->next)
+            {
+                root->next = newlink;
+            }
+            newlink->item = item;
+            return iterator(root, newlink);
+        }
+        iterator insert_after(const_iterator pos, I * const & item)
+        {
+            auto root = static_cast<L *>(this);
+            CHECK_INVALID_ARGUMENT(pos.iter.root == root);
+            CHECK_NULL_POINTER(pos.iter.cur);
+
+            auto link = pos.iter.cur;
+            auto next = link->next;
+            auto newlink = new L();
+            newlink->prev = link;
+            newlink->next = next;
+            link->next = newlink;
+            if (next)
+            {
+                next->prev = newlink;
+            }
+            newlink->item = item;
+            return iterator(root, newlink);
+        }
+        void push_front(I * const & item)
+        {
+            auto root = static_cast<L *>(this);
+            auto link = new L();
+            link->prev = nullptr;
+            if (root->next)
+            {
+                root->next->prev = link;
+                link->next = root->next;
+            }
+            link->item = item;
+            root->next = link;
```

Comment:
Did you notice that push_front could call insert_after?
<pre>
insert_after(iterator(root), item);
</pre>
Only problem is when assigning newlink->prev = link which should be instead null for when link == root